File: solutions/kotlin/resistor-color-trio/1/src/main/kotlin/ResistorColorTrio.kt
object ResistorColorTrio {
    val UNITS = arrayOf("ohms", "kiloohms", "megaohms", "gigaohms")
    fun text(vararg input: Color): String {
        require(input.size >= 3) {
            "Must provide at least 3 colors"
        }
        var trailingZeros = input.get(2).ordinal
        var digits = input.get(0).ordinal * 10 + input.get(1).ordinal
        while (digits % 10 == 0) {
            digits /= 10
            ++trailingZeros
        }
        val unit = UNITS.get(trailingZeros / 3)
        val zeros = "0".repeat(trailingZeros % 3)
        return "$digits$zeros $unit"
    }
}

File: solutions/lua/resistor-color-trio/1/resistor-color-trio.lua
local COLORCODES = {
  black = 0,
  brown = 1,
  red = 2,
  orange = 3,
  yellow = 4,
  green = 5,
  blue = 6,
  violet = 7,
  grey = 8,
  white = 9,
}

local QUANTIFIERS = {
  [0] = "ohms",
  [3] = "kiloohms",
  [6] = "megaohms",
  [9] = "gigaohms",
}

return {
  decode = function(c1, c2, c3)
    for _, c in ipairs({c1, c2, c3}) do
      if type(c) ~= "string" then
        error(string.format([[            
            decode() expects 3 string type inputs.
            received types: %s, %s, %s
            received values: %s, %s, %s
        ]], type(c1), type(c2), type(c3), c1, c2, c3))
      end
      if COLORCODES[c] == nil then
        error(string.format([[            
            decode() could not find the following color.
            color: %s
        ]], c))
      end
    end
    local digits = (10 * COLORCODES[c1] + COLORCODES[c2]) * math.pow(10, COLORCODES[c3] % 3)
    local exponent = COLORCODES[c3] // 3
    while digits >= 1000 do
      digits = digits // 1000
      exponent = exponent + 1
    end
    if QUANTIFIERS[exponent * 3] == nil then
      error(string.format([[          
          input exceeds the valid numeric range.
          exponent of ten: %d
          inputs: %d, %d, %d          
      ]], exponent * 3, COLORCODES[c1], COLORCODES[c2], COLORCODES[c3]))
    end
    return digits, QUANTIFIERS[exponent * 3]
  end
}